{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "extra-newman",
   "metadata": {},
   "source": [
    "# Read related files\n",
    "\n",
    "Non-standardized metadata files\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "divided-handling",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['.git',\n",
       " '.ipynb_checkpoints',\n",
       " 'ACTIONdemonstrator_XOmics_IDs_synthetic.csv',\n",
       " 'ISA_ACTION-Template.ipynb',\n",
       " 'isa_env.yml',\n",
       " 'isa_template',\n",
       " 'LICENSE',\n",
       " 'README.md',\n",
       " 'read_ISA_ACTION-Template.ipynb']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "input_dir = \".\"\n",
    "os.listdir(path = input_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "technical-missouri",
   "metadata": {},
   "source": [
    "## Sample identifiers from ACTION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "valuable-relationship",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data frame dimensions\n",
      " (100, 5)\n",
      "Column names\n",
      " Index(['XOmicsPhenoID', 'XOmicsGenoID', 'XOmicsFamID', 'XOmicsMethylID',\n",
      "       'XOmicsmetaboID'],\n",
      "      dtype='object')\n",
      "Missing value counts\n",
      " XOmicsPhenoID     0\n",
      "XOmicsGenoID      0\n",
      "XOmicsFamID       0\n",
      "XOmicsMethylID    0\n",
      "XOmicsmetaboID    0\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<bound method NDFrame.head of    XOmicsPhenoID XOmicsGenoID XOmicsFamID XOmicsMethylID XOmicsmetaboID\n",
       "0           XOP1         XOG1        XOF1           XOE1          XOM1 \n",
       "1           XOP2         XOG2        XOF2           XOE2           XOM2\n",
       "2           XOP3         XOG3        XOF3           XOE3           XOM3\n",
       "3           XOP4         XOG4        XOF4           XOE4           XOM4\n",
       "4           XOP5         XOG5        XOF5           XOE5           XOM5\n",
       "..           ...          ...         ...            ...            ...\n",
       "95         XOP96        XOG96       XOF96          XOE96          XOM96\n",
       "96         XOP97        XOG97       XOF97          XOE97          XOM97\n",
       "97         XOP98        XOG98       XOF98          XOE98          XOM98\n",
       "98         XOP99        XOG99       XOF99          XOE99          XOM99\n",
       "99        XOP100       XOG100      XOF100         XOE100         XOM100\n",
       "\n",
       "[100 rows x 5 columns]>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IDs_file_path = \"./ACTIONdemonstrator_XOmics_IDs_synthetic.csv\"\n",
    "import pandas as pd\n",
    "IDs_df = pd.read_csv(IDs_file_path)\n",
    "print(\"Data frame dimensions\\n\", IDs_df.shape)\n",
    "print(\"Column names\\n\", IDs_df.columns)\n",
    "print(\"Missing value counts\\n\", IDs_df.isna().sum())\n",
    "IDs_df.head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a666efa",
   "metadata": {},
   "source": [
    "# Create ISA object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c7603953",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/isa_container/volume'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print current directory\n",
    "import numpy as np\n",
    "import os\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "appreciated-maryland",
   "metadata": {},
   "source": [
    "## Investigation\n",
    "\n",
    "Create a new ISA object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "composite-central",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create new investigation with single study\n",
    "from isatools.model import *\n",
    "investigation = Investigation()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "contemporary-threshold",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'X-omics data analysis, integration and stewardship demonstrator dataset: NTR ACTION omics data'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# define title\n",
    "investigation.title = \"X-omics data analysis, integration and stewardship demonstrator dataset: NTR ACTION omics data\"\n",
    "investigation.title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "balanced-working",
   "metadata": {},
   "outputs": [],
   "source": [
    "investigation.description = \"Predict childhood aggression with multi-omics data and demonstrate the FAIRification process and data analysis of a multi-omics project\"\n",
    "investigation.identifier = \"tbd\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "computational-student",
   "metadata": {},
   "source": [
    "## Study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "sticky-correction",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[isatools.model.Study(filename='', identifier='', title='', description='', submission_date='', public_release_date='', contacts=[], design_descriptors=[], publications=[], factors=[], protocols=[], assays=[], sources=[], samples=[], process_sequence=[], other_material=[], characteristic_categories=[], comments=[], units=[])]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add one study to investigation\n",
    "investigation.studies.append(Study())\n",
    "investigation.studies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "competitive-opinion",
   "metadata": {},
   "source": [
    "According to MetaboLights help site, the title should ideally be the same as for a corresponding manuscript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "married-trinity",
   "metadata": {},
   "outputs": [],
   "source": [
    "investigation.studies[0].title = \"X-omics data analysis, integration and stewardship demonstrator dataset: NTR ACTION omics data\"\n",
    "investigation.studies[0].identifier = \"tbd\" # TODO: add identifier; update title\n",
    "investigation.studies[0].filename = \"s_study.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "violent-memorial",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "isatools.model.Study(filename='s_study.txt', identifier='tbd', title='X-omics data analysis, integration and stewardship demonstrator dataset: NTR ACTION omics data', description='', submission_date='', public_release_date='', contacts=[], design_descriptors=[], publications=[], factors=[], protocols=[], assays=[], sources=[], samples=[], process_sequence=[], other_material=[], characteristic_categories=[], comments=[], units=[])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "investigation.studies[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "alien-winter",
   "metadata": {},
   "source": [
    "### Ontologies\n",
    "\n",
    "Ontologies can be searched e.g. at http://www.ontobee.org/ or https://www.ebi.ac.uk/ols/index.\n",
    "\n",
    "FAIR genomes lookups: https://github.com/fairgenomes/fairgenomes-semantic-model/tree/main/lookups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "considered-assault",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ontologies\n",
    "ontologies = {\n",
    "    \"afo\": OntologySource(\n",
    "        name = \"AFO\",\n",
    "        description = \"Allotrope Merged Ontology Suite\"),\n",
    "    \"chebi\": OntologySource(\n",
    "        name = \"CHEBI\",\n",
    "        description = \"Chemical Entities of Biological Interest\"),\n",
    "    \"chmo\": OntologySource(\n",
    "        name = \"CHMO\", \n",
    "        description = \"Chemical Methods Ontology\"),\n",
    "    \"edam\": OntologySource(\n",
    "        name = \"EDAM\", \n",
    "        description = \"Bioinformatics operations, data types, formats, identifiers and topics\"),\n",
    "    \"efo\": OntologySource(\n",
    "        name = \"EFO\", \n",
    "        description = \"Experimental Factor Ontology\"),\n",
    "    \"ero\": OntologySource(\n",
    "        name = \"eagle-i resource ontology\",\n",
    "        description = \"An ontology of research resources such as instruments, protocols, reagents, animal models and biospecimens\"),\n",
    "    \"maxo\": OntologySource(\n",
    "        name = \"MAXO\", \n",
    "        description = \"Medical Action Ontology\"),\n",
    "    \"msio\": OntologySource(\n",
    "        name = \"MSIO\",\n",
    "        description = \"Metabolite Standards Initiative Ontology\"),\n",
    "    \"ncbitaxon\": OntologySource(\n",
    "        name = \"NCBITAXON\", \n",
    "        description = \"NCBI organismal classification\"),\n",
    "    \"ncit\": OntologySource(\n",
    "        name = \"NCIT\", \n",
    "        description = \"NCI Thesaurus OBO Edition\"),\n",
    "    \"obi\": OntologySource(\n",
    "        name = \"OBI\", \n",
    "        description = \"Ontology for Biomedical Investigations\"),\n",
    "    \"pato\": OntologySource(\n",
    "        name = \"PATO\", \n",
    "        description = \"PATO - the Phenotype And Trait Ontology\"),\n",
    "    \"uberon\": OntologySource(\n",
    "        name = \"UBERON\", \n",
    "        description = \"Uber-anatomy ontology\")\n",
    "}\n",
    "# add ontologies to investigation\n",
    "for o in ontologies.values():\n",
    "    investigation.ontology_source_references.append(o)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fatal-johns",
   "metadata": {},
   "source": [
    "### Study protocols"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bronze-offer",
   "metadata": {},
   "source": [
    "Note that the protocol used in the process to derive `sample` from `source` MUST be of type 'sample collection' (see https://isa-specs.readthedocs.io/en/latest/isatab.html#study-table-file). \n",
    "\n",
    "- ISA model source: https://github.com/ISA-tools/isa-api/blob/master/isatools/model.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "personal-customer",
   "metadata": {},
   "outputs": [],
   "source": [
    "study_protocol_params = {\n",
    "    \"anatomical entity\": ProtocolParameter(\n",
    "            parameter_name = OntologyAnnotation(\n",
    "                term = \"anatomical entity\",\n",
    "                term_source = ontologies[\"uberon\"],\n",
    "                term_accession = \"http://purl.obolibrary.org/obo/UBERON_0001062\"))\n",
    "}\n",
    "# define sample collection protocol\n",
    "sample_collection_protocol = Protocol(\n",
    "    name = \"sample collection\", \n",
    "    # see github.com/ISA-tools/isa-specs/blob/master/source/isatab.rst \n",
    "    # -> MUST be of type 'sample collection'\n",
    "    protocol_type = OntologyAnnotation(term = \"sample collection\"),\n",
    "    parameters = [study_protocol_params[\"anatomical entity\"]])\n",
    "investigation.studies[0].protocols.append(sample_collection_protocol)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mysterious-movement",
   "metadata": {},
   "source": [
    "### Study factors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "spare-supervisor",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gender\n",
    "studyfactor_gender = StudyFactor(\n",
    "    name = \"genotypic sex\", \n",
    "    factor_type = OntologyAnnotation( # Ontology source reference\n",
    "        term = \"genotypic sex\", # also used in FAIR genomes\n",
    "        term_source = ontologies[\"pato\"], \n",
    "        term_accession = \"http://purl.obolibrary.org/obo/PATO_0020000\"))\n",
    "# female\n",
    "factorvalue_female = FactorValue(\n",
    "    factor_name = studyfactor_gender, \n",
    "    value = OntologyAnnotation( # str or OntologyAnnotation\n",
    "        term = \"XX Genotype\", # also used in FAIR genomes\n",
    "        term_source = ontologies[\"ncit\"], \n",
    "        term_accession = \"http://purl.obolibrary.org/obo/NCIT_C45976\"))\n",
    "# male\n",
    "factorvalue_male = FactorValue(\n",
    "    factor_name = studyfactor_gender, \n",
    "    value = OntologyAnnotation( # str or OntologyAnnotation\n",
    "        term = \"XY Genotype\", # also used in FAIR genomes\n",
    "        term_source = ontologies[\"ncit\"], \n",
    "        term_accession = \"http://purl.obolibrary.org/obo/NCIT_C45977\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4300e488",
   "metadata": {},
   "source": [
    "## Assays and assay protocols"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "blocked-arctic",
   "metadata": {},
   "source": [
    "### Genotyping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "material-ghost",
   "metadata": {},
   "outputs": [],
   "source": [
    "assay_genotype = Assay(filename = \"a_assay_genotype.txt\",\n",
    "    measurement_type = OntologyAnnotation(term = \"\", term_source = \"\", term_accession = \"\"),\n",
    "    technology_type = OntologyAnnotation(term = \"nucleotide sequencing\", term_source =\"\", term_accession = \"\"),\n",
    "    technology_platform = OntologyAnnotation(term = \"\", term_source = \"\", term_accession = \"\"))\n",
    "# TODO: What sequencing plaform has been used?\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "efficient-tourist",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define extraction and measurement protocols\n",
    "# DNA extraction for both gebotyping and DNA methylation profiling\n",
    "dna_extraction_protocol = Protocol(\n",
    "    name = \"DNA extraction\",\n",
    "    protocol_type = OntologyAnnotation(\n",
    "        term = \"DNA extraction\",\n",
    "        term_source = ontologies[\"obi\"], \n",
    "        term_accession = \"http://purl.obolibrary.org/obo/OBI_0000257\")) \n",
    "# TODO: check type; compare to FAIR genomes\n",
    "investigation.studies[0].protocols.append(dna_extraction_protocol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "through-tactics",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_profiling_protocol = Protocol(\n",
    "    name = \"genotype profiling\",\n",
    "    protocol_type = OntologyAnnotation(\n",
    "        term = \"genotyping\",\n",
    "        term_source = ontologies[\"efo\"],\n",
    "    term_accession = \"http://www.ebi.ac.uk/efo/EFO_0000750\")\n",
    ")\n",
    "investigation.studies[0].protocols.append(genotype_profiling_protocol)\n",
    "# TODO: check type; compare to FAIR genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "devoted-crime",
   "metadata": {},
   "source": [
    "### DNA methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ultimate-clark",
   "metadata": {},
   "outputs": [],
   "source": [
    "assay_methylation = Assay(\n",
    "    filename = \"a_assay_methylation.txt\", \n",
    "    measurement_type = OntologyAnnotation(\n",
    "        term = \"Methylation Beta Value\",\n",
    "        term_source = ontologies[\"ncit\"],\n",
    "        term_accession = \"http://purl.obolibrary.org/obo/NCIT_C164051\"),\n",
    "    technology_type = OntologyAnnotation(\n",
    "        term = \"DNA methylation profiling by array assay\",\n",
    "        term_source = ontologies[\"obi\"], \n",
    "        term_accession = \"http://purl.obolibrary.org/obo/OBI_0001332\"),\n",
    "    technology_platform = OntologyAnnotation(\n",
    "        term = \"Illumina Infinium MethylationEPIC BeadChip\",\n",
    "        term_source = ontologies[\"obi\"], \n",
    "        term_accession = \"http://purl.obolibrary.org/obo/OBI_0002131\")\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cleared-routine",
   "metadata": {},
   "outputs": [],
   "source": [
    "methylation_profiling_protocol = Protocol(\n",
    "    name = \"methylation profiling\",\n",
    "    protocol_type = OntologyAnnotation(\n",
    "        term = \"methylation profiling\",\n",
    "        term_source = ontologies[\"efo\"], \n",
    "        term_accession = \"http://www.ebi.ac.uk/efo/EFO_0000751\"),\n",
    "#    components = [OntologyAnnotation(\n",
    "#        term = \"Illumina Infinium MethylationEPIC BeadChip\",\n",
    "#        term_source = obi, \n",
    "#        term_accession = \"http://purl.obolibrary.org/obo/OBI_0002131\")]\n",
    "    ) \n",
    "investigation.studies[0].protocols.append(methylation_profiling_protocol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "answering-sperm",
   "metadata": {},
   "outputs": [],
   "source": [
    " methylation_data_processing_protocol = Protocol(\n",
    "    name = \"methylation data processing protocol\",\n",
    "    protocol_type = OntologyAnnotation(\n",
    "        term = \"Protocol\",\n",
    "        term_source = ontologies[\"edam\"], \n",
    "        term_accession = \"http://edamontology.org/data_2531\"),\n",
    "    description = \"Sinke, Lucy, van Iterson, Maarten, Cats, Davy, Slieker, Roderick, & Heijmans, Bas. (2019, July 11). DNAmArray: Streamlined workflow for the quality control, normalization, and analysis of Illumina methylation array data (Version 2.1). Zenodo. http://doi.org/10.5281/zenodo.3355292\",\n",
    "    uri = \"http://doi.org/10.5281/zenodo.3355292\")\n",
    "investigation.studies[0].protocols.append(methylation_data_processing_protocol)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unique-iraqi",
   "metadata": {},
   "source": [
    "### Metabolomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "digital-adelaide",
   "metadata": {},
   "outputs": [],
   "source": [
    "assay_metabolomics_amines = Assay(\n",
    "    filename = \"a_assay_metabolomics_amines.txt\",\n",
    "\n",
    "    \n",
    "    measurement_type = OntologyAnnotation(\n",
    "        term = \"targeted metabolite profiling\",\n",
    "        term_source = ontologies[\"msio\"], \n",
    "        term_accession = \"http://purl.obolibrary.org/obo/MSIO_0000100\"),\n",
    "    \n",
    "    technology_type = OntologyAnnotation(\n",
    "        term = \"liquid chromatography-mass spectrometry\",\n",
    "        term_source = ontologies[\"chmo\"],\n",
    "        term_accession = \"http://purl.obolibrary.org/obo/CHMO_0000524\"))\n",
    "    \n",
    "#     sample_type = OntologyAnnotation(\n",
    "#         term = \"urine specimen\",\n",
    "#         term_source = ['obi'],\n",
    "#         term_accession = \"http:/purl.obolibrary.org/obo/OBI_0000651\")\n",
    "    \n",
    "\n",
    "#     \n",
    "#     technology_platform = OntologyAnnotation(\n",
    "#         term = \"\",\n",
    "#         term_source = ontologies[\"\"], \n",
    "#         term_accession = \"\")\n",
    "    # TODO: What exact platform/instrument has been used?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "enormous-wholesale",
   "metadata": {},
   "outputs": [],
   "source": [
    "assay_metabolomics_OA = Assay(\n",
    "    filename = \"a_assay_metabolomics_OA.txt\", \n",
    "    \n",
    "    measurement_type = OntologyAnnotation(\n",
    "        term = \"targeted metabolite profiling\",\n",
    "        term_source = ontologies[\"msio\"], \n",
    "        term_accession = \"http://purl.obolibrary.org/obo/MSIO_0000100\"),\n",
    "    \n",
    "    technology_type = OntologyAnnotation(\n",
    "        term = \"gas chromatography-mass spectrometry\",\n",
    "        term_source = ontologies[\"chmo\"]))\n",
    "        \n",
    "#     sample_type = OntologyAnnotation(\n",
    "#         term = \"urine specimen\",\n",
    "#         term_source = ['obi'],\n",
    "#         term_accession = \"http:/purl.obolibrary.org/obo/OBI_0000651\")\n",
    "    \n",
    "    \n",
    "#     technology_platform = OntologyAnnotation(\n",
    "#         term = \"\",\n",
    "#         term_source = ontologies[\"\"], \n",
    "#         term_accession = \"\")\n",
    "        # TODO: What exact platform/instrument has been used?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "equipped-camel",
   "metadata": {},
   "outputs": [],
   "source": [
    "assay_metabolomics_steroids = Assay(\n",
    "    filename = \"a_assay_metabolomics_steroids.txt\", \n",
    "    \n",
    "    measurement_type = OntologyAnnotation(\n",
    "        term = \"targeted metabolite profiling\",\n",
    "        term_source = ontologies[\"msio\"], \n",
    "        term_accession = \"http://purl.obolibrary.org/obo/MSIO_0000100\"),\n",
    "    \n",
    "    technology_type = OntologyAnnotation(\n",
    "        term = \"high-performance liquid chromatography-mass spectrometry\",\n",
    "        term_source = ontologies[\"chmo\"],\n",
    "        term_accession = \"http://purl.obolibrary.org/obo/CHMO_0000796\"))\n",
    "    \n",
    "#     sample_type = OntologyAnnotation(\n",
    "#         term = \"urine specimen\",\n",
    "#         term_source = ['obi'],\n",
    "#         term_accession = \"http:/purl.obolibrary.org/obo/OBI_0000651\")\n",
    "\n",
    "#     technology_platform = OntologyAnnotation(\n",
    "#         term = \"\",\n",
    "#         term_source = ontologies[\"\"], \n",
    "#         term_accession = \"\")\n",
    "# TODO: What exact platform/instrument has been used?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "excessive-reducing",
   "metadata": {},
   "source": [
    "#### Metabolomics protocols\n",
    "\n",
    "See also https://ebi.ac.uk/metabolights/guides/Protocol/Protocol for protocols required for submission in MetaboLights, i.e. Sample collection, Extraction, Chromatography, Mass spectrometry, Data transformation, and Metabolite identification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "83c3c85c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mtbl_parameters = {\n",
    "    \"Post Extraction\": ProtocolParameter(parameter_name = OntologyAnnotation(term = \"Post Extraction\")),\n",
    "    \"Derivatization\": ProtocolParameter(parameter_name = OntologyAnnotation(term = \"Derivatization\")),\n",
    "    \"Chromatography Instrument\": ProtocolParameter(parameter_name = OntologyAnnotation(term = \"Chromatography Instrument\")),\n",
    "    \"Column model\": ProtocolParameter(parameter_name = OntologyAnnotation(term = \"Column model\")),\n",
    "    \"Column type\": ProtocolParameter(parameter_name = OntologyAnnotation(term = \"Column type\")),\n",
    "    \"Scan polarity\": ProtocolParameter(parameter_name = OntologyAnnotation(term = \"Scan polarity\")), \n",
    "    \"Scan m/z range\": ProtocolParameter(parameter_name = OntologyAnnotation(term = \"Scan m/z range\")), \n",
    "    \"Instrument\": ProtocolParameter(parameter_name = OntologyAnnotation(term = \"Instrument\")), \n",
    "    \"Ion source\": ProtocolParameter(parameter_name = OntologyAnnotation(term = \"Ion source\")), \n",
    "    \"Mass analyzer\": ProtocolParameter(parameter_name = OntologyAnnotation(term = \"Mass analyzer\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "greatest-suggestion",
   "metadata": {},
   "outputs": [],
   "source": [
    "extraction_metabolomics = Protocol(\n",
    "    name = \"Extraction\",\n",
    "    protocol_type = OntologyAnnotation(\n",
    "        term = 'Extraction',\n",
    "        term_source = ontologies[\"ncit\"],\n",
    "        term_accession = 'http://purl.obolibrary.org/obo/NCIT_C61575'),\n",
    "    parameters = [\n",
    "        mtbl_parameters[\"Post Extraction\"],\n",
    "        mtbl_parameters[\"Derivatization\"]\n",
    "    ])\n",
    "investigation.studies[0].protocols.append(extraction_metabolomics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "center-recall",
   "metadata": {},
   "outputs": [],
   "source": [
    "chromatography = Protocol(\n",
    "    name = \"Chromatography\",\n",
    "    protocol_type = OntologyAnnotation(\n",
    "        term = 'Chromatography',\n",
    "        term_source = ontologies[\"ncit\"],\n",
    "        term_accession = 'http://purl.obolibrary.org/obo/NCIT_C16431'),\n",
    "    parameters = [\n",
    "        mtbl_parameters[\"Chromatography Instrument\"],\n",
    "        mtbl_parameters[\"Column model\"],\n",
    "        mtbl_parameters[\"Column type\"]\n",
    "    ])\n",
    "investigation.studies[0].protocols.append(chromatography)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "boolean-poultry",
   "metadata": {},
   "outputs": [],
   "source": [
    "labelling_metabolites = Protocol(\n",
    "    name = \"Labelling metabolites\",\n",
    "    protocol_type = OntologyAnnotation(\n",
    "        term = 'Labelling',\n",
    "        term_source = ontologies[\"chmo\"],\n",
    "        term_accession = 'http://purl.obolibrary.org/obo/CHMO_0001675')\n",
    "    )\n",
    "investigation.studies[0].protocols.append(labelling_metabolites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "responsible-stevens",
   "metadata": {},
   "outputs": [],
   "source": [
    "mass_spectrometry = Protocol(\n",
    "    name = \"Mass spectrometry\",\n",
    "    protocol_type = OntologyAnnotation(\n",
    "        term = 'Mass spectrometry',\n",
    "        term_source = ontologies[\"ncit\"],\n",
    "        term_accession = 'http://purl.obolibrary.org/obo/NCIT_C17156'),\n",
    "    parameters = [\n",
    "        mtbl_parameters[\"Scan polarity\"],\n",
    "        mtbl_parameters[\"Scan m/z range\"],\n",
    "        mtbl_parameters[\"Instrument\"],\n",
    "        mtbl_parameters[\"Ion source\"],\n",
    "        mtbl_parameters[\"Mass analyzer\"]\n",
    "    ])\n",
    "investigation.studies[0].protocols.append(mass_spectrometry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "incorporated-capital",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_transformation = Protocol(\n",
    "    name = \"Data transformation\",\n",
    "    protocol_type = OntologyAnnotation(\n",
    "        term = 'Data Transformation',\n",
    "        term_source = ontologies[\"ncit\"],\n",
    "        term_accession = 'http://purl.obolibrary.org/obo/NCIT_C43582')\n",
    "    )\n",
    "investigation.studies[0].protocols.append(data_transformation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "minimal-miracle",
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolite_identification = Protocol(\n",
    "    name = \"Metabolite identification\",\n",
    "    protocol_type = OntologyAnnotation(\n",
    "        term = 'peak identification',\n",
    "        term_source = ontologies[\"afo\"],\n",
    "        term_accession = 'http://purl.allotrope.erg/ontologies/process#AFP_0003618')\n",
    "    )\n",
    "investigation.studies[0].protocols.append(metabolite_identification)\n",
    "# TODO: Is this the correct ontlogy (source)?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "confirmed-technique",
   "metadata": {},
   "source": [
    "## ACTION samples\n",
    "\n",
    "Add samples to study and link to previously defined protocols and assays.\n",
    "\n",
    "For MetaboLights, sample information should include unique sample name, organism, organism part, sample type (control, QC, experimental sample), other descriptors as factors (age, gender)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "floating-summer",
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx, row in IDs_df.iterrows(): \n",
    "    # add subjects (sources) \n",
    "    # TODO: issue - source should represent a source material such as urine, \n",
    "    #   and sample a respective extract or similar\n",
    "    # check if source was already added already (rows can contain duplicate entries)\n",
    "    #is_new_source = True\n",
    "    source_name = row[\"XOmicsPhenoID\"]\n",
    "    source = next((src for src in investigation.studies[0].sources \n",
    "                   if src.name == source_name), None)\n",
    "    if not source:\n",
    "        #is_new_source = False\n",
    "        # create new source for subject\n",
    "        source = Source(name = row[\"XOmicsPhenoID\"])\n",
    "        # Characteristics - Organism - should be included for Metabolights\n",
    "        # here, organism is defined per source, i.e. individual\n",
    "        source.characteristics.append(\n",
    "            Characteristic(\n",
    "                category = OntologyAnnotation(\n",
    "                    term = \"organism\"), \n",
    "                # TODO: add term source and accession for category? would such information be lost in ISA-Tab?\n",
    "                value = OntologyAnnotation(\n",
    "                    term = \"Homo sapiens\",\n",
    "                    term_source = ontologies[\"ncbitaxon\"],\n",
    "                    term_accession = \"http://purl.bioontology.org/ontology/NCBITAXON/9606\")))\n",
    "        # TODO: check if family ID should/should not be added / is required for analysis\n",
    "        source.characteristics.append(\n",
    "            Characteristic(category = \"family ID\",\n",
    "                           value = row[\"XOmicsFamID\"]))\n",
    "        source.characteristics.append(\n",
    "            Characteristic(\n",
    "                category = OntologyAnnotation( \n",
    "                    term = \"childhood aggressive behaviour measurement\", \n",
    "                    term_source = ontologies[\"efo\"], \n",
    "                    term_accession = \"http://www.ebi.ac.uk/efo/EFO_0007663\"),\n",
    "                value = \"\", #tscore,\n",
    "                unit = OntologyAnnotation( \n",
    "                    term = \"T-score\", \n",
    "                    term_source = ontologies[\"ncit\"], \n",
    "                    term_accession = \"http://purl.obolibrary.org/obo/NCIT_C120401\")))\n",
    "        #source.factor_values.append(FactorValueAggressionScore(value = tscore))\n",
    "        \n",
    "        # add subject to study\n",
    "        investigation.studies[0].sources.append(source)\n",
    "        \n",
    "    # add samples - sample names need to be unique\n",
    "    # urine sample for metabolomics\n",
    "    if not pd.isna(row[\"XOmicsmetaboID\"]):\n",
    "        # check if urine sample was already added to study\n",
    "        urine_sample_name = \"urine_{0}\".format(source_name)\n",
    "        urine_sample = next(\n",
    "            (smpl for smpl in investigation.studies[0].samples \n",
    "             if smpl.name == urine_sample_name), None)\n",
    "        if not urine_sample:\n",
    "            # create a new sample with unique name\n",
    "            urine_sample = Sample(\n",
    "                name = urine_sample_name, \n",
    "                derives_from = [source]) # the individual\n",
    "            # Characteristics - Organism part - should be included for Metabolights\n",
    "            # here, organism part is defined per sample\n",
    "            urine_sample.characteristics.append(\n",
    "                Characteristic(\n",
    "                    category = OntologyAnnotation(term = \"organism part\"),\n",
    "                    value = OntologyAnnotation(\n",
    "                        term = \"urine\",\n",
    "                        term_source = ontologies[\"uberon\"],\n",
    "                        term_accession = \"http://purl.obolibrary.org/obo/UBERON_0001088\")))\n",
    "            # Characteristics - sample type - should be included for Metabolights \n",
    "            # i.e. control, QC, experimental sample\n",
    "            urine_sample.characteristics.append(\n",
    "                Characteristic(\n",
    "                    category = \"sample type\", # TODO: could not find a term yet; sample type is not an ontological term, but required by MetaboLights\n",
    "                    value = OntologyAnnotation(\n",
    "                        term = \"experimental sample\",\n",
    "                        term_source = ontologies[\"chmo\"],\n",
    "                        term_accession = \"http://purl.obolibrary.org/obo/CHMO_0002746\")))\n",
    "            # add urine sample to study\n",
    "            investigation.studies[0].samples.append(urine_sample)\n",
    "            \n",
    "            # check if urine sampling process exists for source\n",
    "            urine_p_name = \"urine_specimen_collection_process_{0}\".format(source.name)\n",
    "            urine_collection_process = next(\n",
    "                (prcs for prcs in investigation.studies[0].process_sequence \n",
    "                 if prcs.name == urine_p_name), None)\n",
    "            if not urine_collection_process:\n",
    "                # define urine sampling process for this subject\n",
    "                urine_collection_process = Process(\n",
    "                    name = urine_p_name, \n",
    "                    executes_protocol = sample_collection_protocol,\n",
    "                    parameter_values = [\n",
    "                        ParameterValue(\n",
    "                            category = study_protocol_params[\"anatomical entity\"], #ProtocolParameter \n",
    "                            value = \"urine\")],\n",
    "                    inputs = [source],\n",
    "                    outputs = [urine_sample])\n",
    "                investigation.studies[0].process_sequence.append(urine_collection_process)\n",
    "            else:\n",
    "                # urine sampling process already exists for the source\n",
    "                # add urine sample to outputs of existing process\n",
    "                urine_collection_process.outputs.append(urine_sample)\n",
    "    \n",
    "    # add samples\n",
    "    # buccal swab sample for genotyping and DNA methylation arrays\n",
    "    if not pd.isna(row[\"XOmicsGenoID\"]) or not pd.isna(row[\"XOmicsMethylID\"]):\n",
    "        buccal_sample_name = \"buccal_mucosa_{0}\".format(source_name)\n",
    "        \n",
    "        buccal_sample = next(\n",
    "            (smpl for smpl in investigation.studies[0].samples \n",
    "             if smpl.name == buccal_sample_name), None)\n",
    "        \n",
    "        if not buccal_sample:\n",
    "            # create sample of buccal mucosa\n",
    "            buccal_sample = Sample(\n",
    "                name = buccal_sample_name, \n",
    "                derives_from = [source]) # same source as urine sample\n",
    "            buccal_sample.characteristics.append(\n",
    "                Characteristic(\n",
    "                    category = OntologyAnnotation(term = \"organism part\"),\n",
    "                    value = \"\"))\n",
    "            # TODO: add more characteristics - compare to urime sample\n",
    "            # add sample to study\n",
    "            investigation.studies[0].samples.append(buccal_sample) \n",
    "            \n",
    "            # check if buccal swab sampling process exists for source\n",
    "            # needs to be checked, because multiple samples can be derived from one source\n",
    "            buccal_p_name = \"buccal_specimen_collection_process_{0}\".format(source.name)\n",
    "            buccal_collection_process = next(\n",
    "                (prcs for prcs in investigation.studies[0].process_sequence \n",
    "                 if prcs.name == buccal_p_name), None)\n",
    "            if not buccal_collection_process:\n",
    "                # define buccal sampling process for this subject\n",
    "                buccal_collection_process = Process(\n",
    "                    name = buccal_p_name, \n",
    "                    executes_protocol = sample_collection_protocol,\n",
    "                    parameter_values = [\n",
    "                        ParameterValue(\n",
    "                            category = study_protocol_params[\"anatomical entity\"], #ProtocolParameter \n",
    "                            value = \"buccal mucosa\")],\n",
    "                    inputs = [source],\n",
    "                    outputs = [buccal_sample]\n",
    "                )\n",
    "                investigation.studies[0].process_sequence.append(buccal_collection_process)\n",
    " \n",
    "            else:\n",
    "                # buccal sampling process already exists for the source\n",
    "                # add buccal sample to outputs of existing process\n",
    "                buccal_collection_process.outputs.append(buccal_sample)\n",
    "                \n",
    "            # NOTE: adding extraction process at study level doesn seem to work\n",
    "            # causes sample to disappear from study file\n",
    "            \n",
    "                \n",
    "\n",
    "        #if not pd.isna(row[\"XOmicsMethylID\"]):\n",
    "        #    dna_extraction_process = next(())\n",
    "        #    assay_methylation.samples.append(buccal_sample) # first check if already added\n",
    "        #    methylation_profiling_process = Process(\n",
    "        #            name = \"methylation_profiling_{0}\".format(source.name),\n",
    "        #            executes_protocol = methylation_profiling_protocol,\n",
    "        #            inputs = [buccal_sample], \n",
    "        ##            outputs = [buccal_dna])\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "decent-budapest",
   "metadata": {},
   "source": [
    "Genotype assay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "honey-salon",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add samples to genotype assay\n",
    "for idx, row in IDs_df.iterrows(): \n",
    "    source_name = row[\"XOmicsPhenoID\"]\n",
    "    if not pd.isna(row[\"XOmicsGenoID\"]):\n",
    "        buccal_sample_name = \"buccal_mucosa_{0}\".format(source_name)\n",
    "        buccal_sample = next(\n",
    "            (smpl for smpl in investigation.studies[0].samples \n",
    "             if smpl.name == buccal_sample_name), None)\n",
    "        genotype_sample = next(\n",
    "            (smpl for smpl in assay_genotype.samples \n",
    "             if smpl.name == buccal_sample_name), None)\n",
    "        if not genotype_sample:\n",
    "            assay_genotype.samples.append(buccal_sample) # first check if already added\n",
    "            # define DNA as material extracted from buccal mucosa sample\n",
    "            # on study level, because the same DNA is used for genotyping AND DNA methylation profiling\n",
    "            # TODO: check if this works; could be that extraction has to be on assay level\n",
    "            # but Study object has process_sequence, i.e. this is technically possible \n",
    "            # define DNA material for this sample\n",
    "            # now trying on assay level\n",
    "            buccal_dna = Material(\n",
    "                name = \"buccal_DNA_{0}\".format(row[\"XOmicsGenoID\"]),\n",
    "                type_ = \"Extract Name\")\n",
    "            # define extraction process for buccal DNA\n",
    "            dna_extraction_process = Process(\n",
    "                name = \"DNA_extraction_{0}\".format(row[\"XOmicsGenoID\"]),\n",
    "                executes_protocol = dna_extraction_protocol,\n",
    "                inputs = [buccal_sample], \n",
    "                outputs = [buccal_dna])\n",
    "            \n",
    "\n",
    "            #if not pd.isna(row[\"XOmicsMethylID\"]):\n",
    "            #    dna_extraction_process = next(())\n",
    "\n",
    "            genotype_profiling_process = Process(\n",
    "                name = \"genotype_profiling_{0}\".format(row[\"XOmicsGenoID\"]),\n",
    "                executes_protocol = genotype_profiling_protocol,\n",
    "                inputs = [buccal_dna])\n",
    "            \n",
    "            plink(dna_extraction_process, genotype_profiling_process)\n",
    "            assay_genotype.process_sequence.append(dna_extraction_process) \n",
    "            assay_genotype.process_sequence.append(genotype_profiling_process) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "vietnamese-newspaper",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add samples to methylation assay\n",
    "for idx, row in IDs_df.iterrows(): \n",
    "    source_name = row[\"XOmicsPhenoID\"]\n",
    "    if not pd.isna(row[\"XOmicsMethylID\"]):\n",
    "        buccal_sample_name = \"buccal_mucosa_{0}\".format(source_name)\n",
    "        buccal_sample = next(\n",
    "            (smpl for smpl in investigation.studies[0].samples \n",
    "             if smpl.name == buccal_sample_name), None)\n",
    "        methylation_sample = next(\n",
    "            (smpl for smpl in assay_methylation.samples \n",
    "             if smpl.name == buccal_sample_name), None)\n",
    "        if not methylation_sample:\n",
    "            assay_methylation.samples.append(buccal_sample) # first check if already added\n",
    "            # define DNA as material extracted from buccal mucosa sample\n",
    "            # on study level, because the same DNA is used for genotyping AND DNA methylation profiling\n",
    "            # TODO: check if this works; could be that extraction has to be on assay level\n",
    "            # but Study object has process_sequence, i.e. this is technically possible \n",
    "            # define DNA material for this sample\n",
    "            # now trying on assay level\n",
    "            buccal_dna = Material(\n",
    "                name = \"buccal_DNA_{0}\".format(row[\"XOmicsMethylID\"]),\n",
    "                type_ = \"Extract Name\")\n",
    "            # define extraction process for buccal DNA\n",
    "            dna_extraction_process = Process(\n",
    "                name = \"DNA_extraction_{0}\".format(row[\"XOmicsMethylID\"]),\n",
    "                executes_protocol = dna_extraction_protocol,\n",
    "                inputs = [buccal_sample], \n",
    "                outputs = [buccal_dna])\n",
    "            \n",
    "\n",
    "            #if not pd.isna(row[\"XOmicsMethylID\"]):\n",
    "            #    dna_extraction_process = next(())\n",
    "\n",
    "            methylation_profiling_process = Process(\n",
    "                name = \"methylation_profiling_{0}\".format(row[\"XOmicsMethylID\"]),\n",
    "                executes_protocol = methylation_profiling_protocol,\n",
    "                inputs = [buccal_dna])\n",
    "            \n",
    "            plink(dna_extraction_process, methylation_profiling_process)\n",
    "            assay_methylation.process_sequence.append(dna_extraction_process) \n",
    "            assay_methylation.process_sequence.append(methylation_profiling_process) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "higher-trinidad",
   "metadata": {},
   "source": [
    "Metabolomics assays"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "residential-reservation",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add samples, processes and datafiles to metabolomics Amines assay\n",
    "\n",
    "Assay = assay_metabolomics_amines\n",
    "\n",
    "# Define datafiles (not all may be relevant)\n",
    "\n",
    "raw_datafile = DataFile(filename=\"link/to/raw/data\", label=\"Raw Spectral Data File\")\n",
    "\n",
    "normalized_datafile = DataFile(filename=\"link/to/normalized_data\", label=\"Normalization Name\")\n",
    "\n",
    "derived_spectral_data_file = DataFile(filename=\"link/to/spectral_file\", label=\"Derived Spectral Data File\")\n",
    "\n",
    "Data_Transformation_Name = DataFile(filename=\"link/to/data_transformation_name\", label=\"Data Transformation Name\")\n",
    "\n",
    "MAF = DataFile(filename=\"link/to/MAF\", label=\"Metabolite Assignment File\")\n",
    "  \n",
    "\n",
    "# Loop over samples and add process to samples\n",
    "for idx, row in IDs_df.iterrows():\n",
    "    source_name = row[\"XOmicsPhenoID\"]\n",
    "#     print(source_name)\n",
    "    if not pd.isna(row[\"XOmicsmetaboID\"]):    \n",
    "        #         print(row['XOmicsmetaboID'])\n",
    "        urine_sample_name = \"urine_{0}\".format(source_name)\n",
    "#         print(urine_sample_name)\n",
    "        urine_sample = next(\n",
    "            (smpl for smpl in investigation.studies[0].samples \n",
    "             if smpl.name == urine_sample_name), None)\n",
    "\n",
    "        metabolomics_sample = next(\n",
    "            (smpl for smpl in Assay.samples \n",
    "             if smpl.name == urine_sample_name), None)\n",
    "        \n",
    "        if not metabolomics_sample:\n",
    "            \n",
    "            \n",
    "            \n",
    "            ## Extraction\n",
    "            Post_extraction = ParameterValue(\n",
    "                category = mtbl_parameters[\"Post Extraction\"], \n",
    "                value = OntologyAnnotation(term = \"1 uL borate buffer (pH 8.8) with AQC reagent\"))\n",
    "            Derivatization = ParameterValue(\n",
    "                category = mtbl_parameters[\"Derivatization\"], \n",
    "                value = \"AQC\")\n",
    "            \n",
    "            material_extract = Material(\n",
    "                name = \"extract_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                type_ = \"Extract Name\")\n",
    "            \n",
    "            extraction_process = Process(\n",
    "                executes_protocol=extraction_metabolomics, \n",
    "                parameter_values=[Post_extraction, Derivatization],\n",
    "                inputs = [urine_sample],\n",
    "                outputs = [material_extract])\n",
    "            \n",
    "            \n",
    "            ## Labeling\n",
    "            material_label = Material(\n",
    "                name =\"labeled_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                type_ =\"Labeled Extract Name\")\n",
    "            \n",
    "            labelling_process = Process(\n",
    "                executes_protocol=labelling_metabolites,\n",
    "                inputs = [extraction_process.outputs[0]],\n",
    "                outputs = [material_label])\n",
    "            \n",
    "            \n",
    "            ## Chromatography\n",
    "#             separated_molecules = Material(\n",
    "#                 name = \"separated_molecules_{0}\".format(row[\"XOmicsmetaboID\"],\n",
    "#                 type_ =\"Labeled Extract Name\")\n",
    "#             )\n",
    "            \n",
    "            instrument = ParameterValue(\n",
    "                category = mtbl_parameters[\"Chromatography Instrument\"], \n",
    "                value = \"Agilent 1290 Infinity II\")\n",
    "            column_model = ParameterValue(\n",
    "                category = mtbl_parameters[\"Column model\"], \n",
    "                value = \"Accq-Tag Ultra column (waters + FURHTER SPECS?)\")\n",
    "            column_type = ParameterValue(\n",
    "                category = mtbl_parameters[\"Column type\"], \n",
    "                value = \"reverse phase\")\n",
    "\n",
    "            chromatography_process = Process(\n",
    "                name = \"chromatography_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol = chromatography,\n",
    "                parameter_values = [instrument, column_model, column_type],\n",
    "                inputs = [labelling_process.outputs[0]],\n",
    "                outputs = []\n",
    "#                 outputs = [separated_molecules]\n",
    "            )\n",
    "            \n",
    "            \n",
    "            ## Mass spectrometry\n",
    "            scan_polarity = ParameterValue(\n",
    "                category = mtbl_parameters[\"Scan polarity\"], \n",
    "                value = \"positive\")\n",
    "            scan_range = ParameterValue(\n",
    "                category = mtbl_parameters[\"Scan m/z range\"], \n",
    "                value = \"5-2000?\")\n",
    "            instrument = ParameterValue(\n",
    "                category = mtbl_parameters[\"Instrument\"], \n",
    "                value = \"AB SCIEX Qtrap 6500\")\n",
    "            ion_source = ParameterValue(\n",
    "                category = mtbl_parameters[\"Ion source\"], \n",
    "                value = \"ESI\")\n",
    "            mass_analyzer = ParameterValue(\n",
    "                category = mtbl_parameters[\"Mass analyzer\"], \n",
    "                value = \"triple quadrupole linear ion trap\")\n",
    "            \n",
    "            mass_spectrometry_process = Process(\n",
    "                name = \"mass_spectrometry_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol= mass_spectrometry,\n",
    "                parameter_values = [scan_polarity, scan_range, instrument, ion_source, mass_analyzer],\n",
    "#                 inputs = [separated_molecules],\n",
    "                inputs = [],\n",
    "                outputs = [raw_datafile]\n",
    "            )\n",
    "            \n",
    "            \n",
    "           ## Data transformation\n",
    "            data_transformation_process = Process(\n",
    "                name = \"data_transformation_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol = data_transformation,\n",
    "                inputs = [raw_datafile],\n",
    "                outputs  = [normalized_datafile, derived_spectral_data_file]\n",
    "            )\n",
    "            \n",
    "            \n",
    "            ## Metabolite identification\n",
    "            metabolite_identification_process = Process(\n",
    "                name = \"metabolite_identification_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol =  metabolite_identification,\n",
    "                inputs = [normalized_datafile],\n",
    "                outputs= [Data_Transformation_Name, MAF]\n",
    "            )\n",
    "            \n",
    "            \n",
    "            # Link processes\n",
    "            plink(extraction_process, labelling_process)\n",
    "            plink(labelling_process, chromatography_process)\n",
    "            plink(chromatography_process, mass_spectrometry_process)\n",
    "            plink(mass_spectrometry_process, data_transformation_process)\n",
    "            plink(data_transformation_process, metabolite_identification_process)\n",
    "            \n",
    "            \n",
    "            # Add samples, materials and data files to the amines assay\n",
    "            Assay.samples.append(urine_sample)\n",
    "            Assay.other_material.append(material_extract)\n",
    "            Assay.other_material.append(material_label)\n",
    "#             Assay.other_material.append(separated_molecules)\n",
    "            Assay.data_files.append(raw_datafile)\n",
    "            Assay.data_files.append(normalized_datafile)\n",
    "            Assay.data_files.append(derived_spectral_data_file)\n",
    "            Assay.data_files.append(Data_Transformation_Name)                                                                                                   \n",
    "            Assay.data_files.append(MAF)\n",
    "            \n",
    "            \n",
    "            ## Add processes to the amines assay\n",
    "            Assay.process_sequence.append(extraction_process)\n",
    "            Assay.process_sequence.append(labelling_process)\n",
    "            Assay.process_sequence.append(chromatography_process)\n",
    "            Assay.process_sequence.append(mass_spectrometry_process)\n",
    "            Assay.process_sequence.append(data_transformation_process)\n",
    "            Assay.process_sequence.append(metabolite_identification_process)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "informal-paragraph",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add samples, processes and datafiles to metabolomics OA assay\n",
    "\n",
    "Assay = assay_metabolomics_OA\n",
    "\n",
    "# Define datafiles (not all may be relevant)\n",
    "\n",
    "raw_datafile = DataFile(filename=\"link/to/raw/data\", label=\"Raw Spectral Data File\")\n",
    "\n",
    "normalized_datafile = DataFile(filename=\"link/to/normalized_data\", label=\"Normalization Name\")\n",
    "\n",
    "derived_spectral_data_file = DataFile(filename=\"link/to/spectral_file\", label=\"Derived Spectral Data File\")\n",
    "\n",
    "Data_Transformation_Name = DataFile(filename=\"link/to/data_transformation_name\", label=\"Data Transformation Name\")\n",
    "\n",
    "MAF = DataFile(filename=\"link/to/MAF\", label=\"Metabolite Assignment File\")\n",
    "  \n",
    "\n",
    "# Loop over samples and add process to samples\n",
    "for idx, row in IDs_df.iterrows():\n",
    "    source_name = row[\"XOmicsPhenoID\"]\n",
    "#     print(source_name)\n",
    "    if not pd.isna(row[\"XOmicsmetaboID\"]):\n",
    "#         print(row['XOmicsmetaboID'])\n",
    "        urine_sample_name = \"urine_{0}\".format(source_name)\n",
    "#         print(urine_sample_name)\n",
    "        urine_sample = next(\n",
    "            (smpl for smpl in investigation.studies[0].samples \n",
    "             if smpl.name == urine_sample_name), None)\n",
    "\n",
    "        metabolomics_sample = next(\n",
    "            (smpl for smpl in Assay.samples \n",
    "             if smpl.name == urine_sample_name), None)\n",
    "        \n",
    "        if not metabolomics_sample:\n",
    "            Assay.samples.append(urine_sample)\n",
    "            \n",
    "            \n",
    "            ## Extraction\n",
    "            Post_extraction = ParameterValue(\n",
    "                category = mtbl_parameters[\"Post Extraction\"], \n",
    "                value = \"1 uL pyridine\")\n",
    "            Derivatization = ParameterValue(\n",
    "                category = mtbl_parameters[\"Derivatization\"], \n",
    "                value = \"oximation followed by silylation\")\n",
    "            \n",
    "       \n",
    "            \n",
    "            material_extract = Material(\n",
    "                name = \"extract_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                type_ = \"Extract Name\")\n",
    "                \n",
    "            extraction_process = Process(\n",
    "                executes_protocol=extraction_metabolomics, \n",
    "                parameter_values=[Post_extraction, Derivatization],\n",
    "                inputs = [urine_sample],\n",
    "                outputs = [material_extract])\n",
    "            \n",
    "            \n",
    "            ## Labeling\n",
    "            material_label = Material(\n",
    "                name =\"labeled_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                type_ =\"Labeled Extract Name\")\n",
    "\n",
    "            \n",
    "            labelling_process = Process(\n",
    "                executes_protocol=labelling_metabolites,\n",
    "                inputs = [extraction_process.outputs[0]],\n",
    "                outputs = [material_label])\n",
    "            \n",
    "            \n",
    "#             ## Chromatography\n",
    "#             separated_molecules = Material(\n",
    "#                 name = \"separated_molecules_{0}\".format(row[\"XOmicsmetaboID\"],\n",
    "#                 type_ =\"Labeled Extract Name\")\n",
    "#             )\n",
    "            \n",
    "            instrument = ParameterValue(\n",
    "                category = mtbl_parameters[\"Chromatography Instrument\"], \n",
    "                value = \"Agilent Technologies 7890A\")\n",
    "            column_model = ParameterValue(\n",
    "                category = mtbl_parameters[\"Column model\"], \n",
    "                value = \"HP-5MS UI (5% Phenyl Methyl Silox), 30 m x 0.25 m ID column with a film thickness of 25 um\")\n",
    "            column_type = ParameterValue(\n",
    "                category = mtbl_parameters[\"Column type\"], \n",
    "                value = \"low polarity\")\n",
    "\n",
    "            \n",
    "            chromatography_process = Process(\n",
    "                name = \"chromatography_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol = chromatography,\n",
    "                parameter_values = [instrument, column_model, column_type],\n",
    "                inputs = [labelling_process.outputs[0]], \n",
    "                outputs = [],\n",
    "#                 outputs = [separated_molecules]\n",
    "            )\n",
    "            \n",
    "            ## Mass spectrometry\n",
    "            scan_polarity = ParameterValue(\n",
    "                category = mtbl_parameters[\"Scan polarity\"], \n",
    "                value = \"positive\")\n",
    "            scan_range = ParameterValue(\n",
    "                category = mtbl_parameters[\"Scan m/z range\"], \n",
    "                value = \"50-500\")\n",
    "            instrument = ParameterValue(\n",
    "                category = mtbl_parameters[\"Instrument\"], \n",
    "                value = \"Agilent Technologies mass selective detector (MSD 5975C) and MultiPurpose Sampler (MPS, MXY016-02A, GERSTEL)\")\n",
    "            ion_source = ParameterValue(\n",
    "                category = mtbl_parameters[\"Ion source\"], \n",
    "                value = \"EI (70 eV)\")\n",
    "            mass_analyzer = ParameterValue(\n",
    "                category = mtbl_parameters[\"Mass analyzer\"], \n",
    "                value = \"single-quadrupole\")\n",
    "            \n",
    "            \n",
    "            mass_spectrometry_process = Process(\n",
    "                name = \"mass_spectrometry_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol= mass_spectrometry,\n",
    "                parameter_values = [scan_polarity, scan_range, instrument, ion_source, mass_analyzer],\n",
    "                inputs = [],\n",
    "#                 inputs = [separated_molecules],\n",
    "                outputs = [raw_datafile]\n",
    "            )\n",
    "            \n",
    "            \n",
    "           ## Data transformation\n",
    "            data_transformation_process = Process(\n",
    "                name = \"data_transformation_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol = data_transformation,\n",
    "                inputs = [raw_datafile],\n",
    "                outputs  = [normalized_datafile, derived_spectral_data_file]\n",
    "            )\n",
    "            \n",
    "            ## Metabolite identification\n",
    "            metabolite_identification_process = Process(\n",
    "                name = \"metabolite_identification_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol =  metabolite_identification,\n",
    "                inputs = [normalized_datafile],\n",
    "                outputs= [Data_Transformation_Name, MAF]\n",
    "            )\n",
    "            \n",
    "            ## Link processes\n",
    "            plink(extraction_process, labelling_process)\n",
    "            plink(labelling_process, chromatography_process)\n",
    "            plink(chromatography_process, mass_spectrometry_process)\n",
    "            plink(mass_spectrometry_process, data_transformation_process)\n",
    "            plink(data_transformation_process, metabolite_identification_process)\n",
    "            \n",
    "            ## Add samples, materials and data files to the OA assay\n",
    "            Assay.other_material.append(material_extract)\n",
    "            Assay.other_material.append(material_label)\n",
    "#             Assay.other_material.append(separated_molecules)\n",
    "            Assay.data_files.append(raw_datafile)\n",
    "            Assay.data_files.append(normalized_datafile)\n",
    "            Assay.data_files.append(derived_spectral_data_file)\n",
    "            Assay.data_files.append(Data_Transformation_Name)                                                                                                   \n",
    "            Assay.data_files.append(MAF)\n",
    "            \n",
    "            ## Add processes to the OA assay\n",
    "            Assay.process_sequence.append(extraction_process)\n",
    "            Assay.process_sequence.append(labelling_process)\n",
    "            Assay.process_sequence.append(chromatography_process)\n",
    "            Assay.process_sequence.append(mass_spectrometry_process)\n",
    "            Assay.process_sequence.append(data_transformation_process)\n",
    "            Assay.process_sequence.append(metabolite_identification_process)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "silent-fever",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add samples, processes and datafiles to metabolomics steroids assay\n",
    "\n",
    "Assay = assay_metabolomics_steroids\n",
    "\n",
    "# Define datafiles (not all may be relevant)\n",
    "\n",
    "raw_datafile = DataFile(filename=\"link/to/raw/data\", label=\"Raw Spectral Data File\")\n",
    "\n",
    "normalized_datafile = DataFile(filename=\"link/to/normalized_data\", label=\"Normalization Name\")\n",
    "\n",
    "derived_spectral_data_file = DataFile(filename=\"link/to/spectral_file\", label=\"Derived Spectral Data File\")\n",
    "\n",
    "Data_Transformation_Name = DataFile(filename=\"link/to/data_transformation_name\", label=\"Data Transformation Name\")\n",
    "\n",
    "MAF = DataFile(filename=\"link/to/MAF\", label=\"Metabolite Assignment File\")\n",
    "  \n",
    "\n",
    "# Loop over samples and add process to samples\n",
    "for idx, row in IDs_df.iterrows():\n",
    "    source_name = row[\"XOmicsPhenoID\"]\n",
    "#     print(source_name)\n",
    "    if not pd.isna(row[\"XOmicsmetaboID\"]):\n",
    "#         print(row['XOmicsmetaboID'])\n",
    "        urine_sample_name = \"urine_{0}\".format(source_name)\n",
    "#         print(urine_sample_name)\n",
    "        urine_sample = next(\n",
    "            (smpl for smpl in investigation.studies[0].samples \n",
    "             if smpl.name == urine_sample_name), None)\n",
    "\n",
    "        metabolomics_sample = next(\n",
    "            (smpl for smpl in Assay.samples \n",
    "             if smpl.name == urine_sample_name), None)\n",
    "        \n",
    "        if not metabolomics_sample:\n",
    "            Assay.samples.append(urine_sample)\n",
    "            \n",
    "            \n",
    "            ## Extraction\n",
    "            Post_extraction = ParameterValue(\n",
    "                category = mtbl_parameters[\"Post Extraction\"], \n",
    "                value = \"1 uL filtered urine\")\n",
    "            Derivatization = ParameterValue(\n",
    "                category = mtbl_parameters[\"Derivatization\"], \n",
    "                value = \"NA\")\n",
    "\n",
    "            \n",
    "            material_extract = Material(\n",
    "                name = \"extract_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                type_ = \"Extract Name\")\n",
    "            \n",
    "            extraction_process = Process(\n",
    "                executes_protocol = extraction_metabolomics, \n",
    "                parameter_values = [Post_extraction, Derivatization],\n",
    "                inputs = [urine_sample],\n",
    "                outputs = [material_extract])\n",
    "            \n",
    "            \n",
    "            ## Labeling\n",
    "            material_label = Material(\n",
    "                name =\"labeled_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                type_ =\"Labeled Extract Name\")\n",
    "\n",
    "            \n",
    "            labelling_process = Process(\n",
    "                executes_protocol=labelling_metabolites,\n",
    "                inputs = [extraction_process.outputs[0]],\n",
    "                outputs = [material_label])\n",
    "            \n",
    "            \n",
    "            ## Chromatography\n",
    "#             separated_molecules = Material(\n",
    "#                 name = \"new_separated_molecules_{0}\".format(row[\"XOmicsmetaboID\"],\n",
    "#                 type_ =\"Labeled Extract Name\")  \n",
    "#             )\n",
    "            \n",
    "            instrument = ParameterValue(\n",
    "                category = mtbl_parameters[\"Chromatography Instrument\"], \n",
    "                value = \"Agilent 1290\")\n",
    "            column_model = ParameterValue(\n",
    "                category = mtbl_parameters[\"Column model\"], \n",
    "                value = \"Acquity UPLC CSH C18 column (Waters)\")\n",
    "            column_type = ParameterValue(\n",
    "                category = mtbl_parameters[\"Column type\"], \n",
    "                value = \"reverse phase\")\n",
    "\n",
    "            \n",
    "            chromatography_process = Process(\n",
    "                name = \"chromatography_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol = chromatography,\n",
    "                parameter_values = [instrument, column_model, column_type],\n",
    "                inputs = [labelling_process.outputs[0]], \n",
    "                outputs = []\n",
    "                #outputs = [separated_molecules]\n",
    "            )\n",
    "            \n",
    "            ## Mass spectrometry\n",
    "            scan_polarity = ParameterValue(\n",
    "                category = mtbl_parameters[\"Scan polarity\"], \n",
    "                value = \"switching positive and negative ion mode !! MAYBE SERPARATE INTO NEGATIVE AND POSITIVE ASSAY?\")\n",
    "            scan_range = ParameterValue(\n",
    "                category = mtbl_parameters[\"Scan m/z range\"], \n",
    "                value = \"5-3000?\")\n",
    "            instrument = ParameterValue(\n",
    "                category = mtbl_parameters[\"Instrument\"], \n",
    "                value = \"Agilent 6460\")\n",
    "            ion_source = ParameterValue(\n",
    "                category = mtbl_parameters[\"Ion source\"], \n",
    "                value = \"ESI\")\n",
    "            mass_analyzer = ParameterValue(\n",
    "                category = mtbl_parameters[\"Mass analyzer\"], \n",
    "                value = \"triple quadrupole\")\n",
    "            \n",
    "            \n",
    "            mass_spectrometry_process = Process(\n",
    "                name = \"mass_spectrometry_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol= mass_spectrometry,\n",
    "                parameter_values = [scan_polarity, scan_range, instrument, ion_source, mass_analyzer],\n",
    "               # inputs = [separated_molecules],\n",
    "                inputs = [],\n",
    "                outputs = [raw_datafile]\n",
    "            )\n",
    "            \n",
    "            \n",
    "           ## Data transformation\n",
    "            data_transformation_process = Process(\n",
    "                name = \"data_transformation_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol = data_transformation,\n",
    "                inputs = [raw_datafile],\n",
    "                outputs  = [normalized_datafile, derived_spectral_data_file]\n",
    "            )\n",
    "            \n",
    "            ## Metabolite identification\n",
    "            metabolite_identification_process = Process(\n",
    "                name = \"metabolite_identification_{0}\".format(row[\"XOmicsmetaboID\"]),\n",
    "                executes_protocol =  metabolite_identification,\n",
    "                inputs = [normalized_datafile],\n",
    "                outputs= [Data_Transformation_Name, MAF]\n",
    "            )\n",
    "            \n",
    "            ## Link processes\n",
    "            plink(extraction_process, labelling_process)\n",
    "            plink(labelling_process, chromatography_process)\n",
    "            plink(chromatography_process, mass_spectrometry_process)\n",
    "            plink(mass_spectrometry_process, data_transformation_process)\n",
    "            plink(data_transformation_process, metabolite_identification_process)\n",
    "            \n",
    "            ## Add samples, materials and data files to the steroids assay\n",
    "            Assay.other_material.append(material_extract)\n",
    "            Assay.other_material.append(material_label)\n",
    "            # Assay.other_material.append(separated_molecules)\n",
    "            Assay.data_files.append(raw_datafile)\n",
    "            Assay.data_files.append(normalized_datafile)\n",
    "            Assay.data_files.append(derived_spectral_data_file)\n",
    "            Assay.data_files.append(Data_Transformation_Name)                                                                                                   \n",
    "            Assay.data_files.append(MAF)\n",
    "            \n",
    "            ## Add processes to the steroids assay\n",
    "            Assay.process_sequence.append(extraction_process)\n",
    "            Assay.process_sequence.append(labelling_process)\n",
    "            Assay.process_sequence.append(chromatography_process)\n",
    "            Assay.process_sequence.append(mass_spectrometry_process)\n",
    "            Assay.process_sequence.append(data_transformation_process)\n",
    "            Assay.process_sequence.append(metabolite_identification_process)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "given-beaver",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add assays to study\n",
    "investigation.studies[0].assays.append(assay_genotype)\n",
    "investigation.studies[0].assays.append(assay_methylation)\n",
    "investigation.studies[0].assays.append(assay_metabolomics_amines)\n",
    "investigation.studies[0].assays.append(assay_metabolomics_OA)\n",
    "investigation.studies[0].assays.append(assay_metabolomics_steroids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dramatic-webcam",
   "metadata": {},
   "source": [
    "# Write ISA-Tab files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "regular-palmer",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# create ISA files directory \n",
    "out_dir = \"isa_template\"\n",
    "if not os.path.isdir(out_dir):\n",
    "    os.makedirs(out_dir)\n",
    "# write to ISA-Tab\n",
    "from isatools import isatab\n",
    "isatab.dump(investigation, out_dir)\n",
    "print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "welcome-kitty",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "NotebookApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "print(json.dumps(investigation, cls=ISAJSONEncoder, sort_keys=True, indent=4, separators=(',', ': ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "80b5d3ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from isatools.isajson import ISAJSONEncoder\n",
    "with open (os.path.join(out_dir, 'isa.json'), \"w\") as out_file:\n",
    "    json.dump(\n",
    "        investigation,\n",
    "        out_file,\n",
    "        cls = ISAJSONEncoder,\n",
    "        sort_keys = True,\n",
    "        indent = 4,\n",
    "        separators = (',', ': ')\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:isa_env] *",
   "language": "python",
   "name": "conda-env-isa_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
